### PR TITLE
EVG-6178: refresh merge task dependencies for generate.task

### PR DESCRIPTION
**Pull request by ybrill**
The CLI commit queue adds a merge task to the patch testing the changes and sets its dependencies to all other tasks configured for the commit queue. Since this happens while the task is finalized there was no way for the merge to depend on generated tasks.

This PR is to add generated tasks matching the commit queue task definitions to the merge task's dependencies. Since the generating task doesn't go green until it's finished generating tasks, a merge can be made conditional on the generated tasks passing by specifying both the generating task and the generated tasks as commit queue tasks.

### CODE REVIEW

File: model/generate.go [hunk 1]
```diff
@@ -392,3 +398,21 @@ func (g *GeneratedProject) validateGeneratedProject(p *Project, cachedProject pr
 
 	return errors.WithStack(catcher.Resolve())
 }
+
+func updateMergeTaskDependencies(p *Project, v *Version) error {
```
Comment (tychoish) [inline model/generate.go]: should this be a method on version?

File: model/generate.go [hunk 2]
```diff
@@ -392,3 +398,21 @@ func (g *GeneratedProject) validateGeneratedProject(p *Project, cachedProject pr
...
+	execTaskIDTable := NewTaskIdTable(p, v, "", "").ExecutionTasks
+	mergeTaskDependencies := make([]task.Dependency, 0, len(execPairs))
+	for _, pair := range execPairs {
+		depID := execTaskIDTable.GetId(pair.Variant, pair.TaskName)
+		mergeTaskDependencies = append(mergeTaskDependencies, task.Dependency{TaskId: depID, Status: evergreen.TaskSucceeded})
+	}
+
+	// update task in the database
+	mergeTaskID := execTaskIDTable.GetId(evergreen.MergeTaskVariant, evergreen.MergeTaskName)
+	return task.UpdateOne(bson.M{task.IdKey: mergeTaskID}, bson.M{"$set": bson.M{task.DependsOnKey: mergeTaskDependencies}})
```
Comment (tychoish) [inline model/generate.go]: combine previous two statments, maybe on multiple lines and wrap? 

maybe it makes sense to write a function for this specific update operation as a function in the task package?

File: model/generate.go [hunk 3]
```diff
@@ -392,3 +398,21 @@ func (g *GeneratedProject) validateGeneratedProject(p *Project, cachedProject pr
...
+	execPairs, _, err := p.BuildProjectTVPairsWithAlias(evergreen.CommitQueueAlias)
+	if err != nil {
+		return errors.Wrap(err, "can't get alias pairs")
+	}
+
+	execTaskIDTable := NewTaskIdTable(p, v, "", "").ExecutionTasks
+	mergeTaskDependencies := make([]task.Dependency, 0, len(execPairs))
+	for _, pair := range execPairs {
+		depID := execTaskIDTable.GetId(pair.Variant, pair.TaskName)
+		mergeTaskDependencies = append(mergeTaskDependencies, task.Dependency{TaskId: depID, Status: evergreen.TaskSucceeded})
```
Comment (tychoish) [inline model/generate.go]: I'd combine these lines, but I suspect it doesn't matter much.

Comment (ybrill) [reply model/generate.go]: done.

File: model/generate.go (hunk 2 above)
Comment (ybrill) [reply model/generate.go]: done and done.

File: model/generate.go (hunk 1 above)
Comment (ybrill) [reply model/generate.go]: done.

File: model/version.go [hunk 4]
```diff
@@ -211,6 +211,33 @@ func VersionGetHistory(versionId string, N int) ([]Version, error) {
 	return versions, nil
 }
 
+func (v *Version) UpdateMergeTaskDependencies(p *Project) error {
```
Comment (bsamek) [inline model/version.go]: Could this race if something else interweaves its dependency updates? Could you use an atomic update, and error (maybe with a very high error level, so we get a slack message) if this happens?
```
	return UpdateOne(
		bson.M{
			IdKey: t.Id,
                         DependsOnKey: <previous state>
		},
		bson.M{
			"$set": bson.M{DependsOnKey: dependsOn},
		},
	)
}
<handle update error>
```

Comment (ybrill) [reply model/version.go]: done.

**Comment by ybrill** [comment]
Made the check for whether a task is a commit queue merge more robust (to address the concern: what if a user has a task with the same name?)

**Pull request merged**